File: adapters/express/src/gateway-session.js
"use strict";

const crypto = require("crypto");

/**
 * In-memory session store for LLM Gateway dialogue sessions.
 * Sessions auto-expire after TTL (default 5 minutes).
 */
class GatewaySessionStore {
  constructor(ttlMs = 300_000) {
    this._ttl = ttlMs;
    this._store = new Map();
    this._cleanupInterval = setInterval(() => this._sweep(), 60_000);
    if (this._cleanupInterval.unref) this._cleanupInterval.unref();
  }

  /**
   * Create a new session with the given data.
   * @param {object} data - Context data to store
   * @returns {string} 16-char hex session ID prefixed with "gw_"
   */
  create(data) {
    const id = "gw_" + crypto.randomBytes(8).toString("hex");
    this._store.set(id, { data, created: Date.now() });
    return id;
  }

  /**
   * Retrieve session data by ID. Returns null if expired or not found.
   * @param {string} id
   * @returns {object|null}
   */
  get(id) {
    const entry = this._store.get(id);
    if (!entry) return null;
    if (Date.now() - entry.created > this._ttl) {
      this._store.delete(id);
      return null;
    }
    return entry.data;
  }

  /**
   * Delete a session (cleanup after use).
   * @param {string} id
   */
  delete(id) {
    this._store.delete(id);
  }

  /**
   * Remove all expired sessions from the store.
   */
  _sweep() {
    const now = Date.now();
    for (const [id, entry] of this._store) {
      if (now - entry.created > this._ttl) {
        this._store.delete(id);
      }
    }
  }
}

module.exports = { GatewaySessionStore };

File: adapters/fastapi/openfeeder_fastapi/gateway_session.py
"""
OpenFeeder FastAPI Adapter — Gateway Session Store

In-memory session store for LLM Gateway dialogue sessions.
Sessions auto-expire after TTL (default 5 minutes).
"""

from __future__ import annotations

import secrets
import time
import threading
from typing import Any


class GatewaySessionStore:
    """Thread-safe in-memory session store with TTL expiry."""

    def __init__(self, ttl_seconds: int = 300) -> None:
        self._ttl = ttl_seconds
        self._store: dict[str, dict[str, Any]] = {}
        self._lock = threading.Lock()
        # Background sweep every 60s
        self._timer = threading.Timer(60.0, self._sweep)
        self._timer.daemon = True
        self._timer.start()

    def create(self, data: dict[str, Any]) -> str:
        """Create a new session. Returns 'gw_' prefixed hex ID."""
        session_id = "gw_" + secrets.token_hex(8)
        with self._lock:
            self._store[session_id] = {"data": data, "created": time.time()}
        return session_id

    def get(self, session_id: str) -> dict[str, Any] | None:
        """Retrieve session data by ID. Returns None if expired or not found."""
        with self._lock:
            entry = self._store.get(session_id)
            if entry is None:
                return None
            if time.time() - entry["created"] > self._ttl:
                del self._store[session_id]
                return None
            return entry["data"]

    def delete(self, session_id: str) -> None:
        """Delete a session after use."""
        with self._lock:
            self._store.pop(session_id, None)

    def _sweep(self) -> None:
        """Remove expired sessions."""
        now = time.time()
        with self._lock:
            expired = [
                sid for sid, entry in self._store.items()
                if now - entry["created"] > self._ttl
            ]
            for sid in expired:
                del self._store[sid]
        # Reschedule
        self._timer = threading.Timer(60.0, self._sweep)
        self._timer.daemon = True
        self._timer.start()
